### PR TITLE
Use repo.postsync.d instead of postsync.d

### PR DESCRIPTION
**Pull request by redneb**
This is related to #6.

This PR moves the postsync hooks to repo.postsync.d. This has the advantage that the scripts become independent of [gentoo bug #558322](https://bugs.gentoo.org/show_bug.cgi?id=558322) and that we do not need to consult the `PORTDIR` variable (which I believe is now deprecated) since the repo.postsync.d scripts are invoked with the path to repository as their third command line argument. Additionally, we can now get rid of the `etc/portage/check-portdir.sh` library script.

I split this into 3 commits to make it easier to review. The first 2 commits should be fairly uncontroversial. The third one changes the semantics of the metadata cache generation.

I thought about the cache for a while and I think it's better if this config scripts for portage only affect the cache of the `gentoo` repository. After all, this is a config about letting portage use git to sync the gentoo repository; why should it affect the behavior of the other repositories? Users will not expect that. Besides, if someone wants to update the cache of the other repos, they can just activate the `/etc/portage/repo.postsync.d/example` script that comes with portage; it does exactly that.

If you don't like this approach vis-a-vis the cache, you can still take the first 2 commits.

### CODE REVIEW

File: etc/portage/repo.postsync.d/sync_gentoo_cache [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+# Number of jobs for egencache, default is number or processors.
+parallel_jobs="$(nproc)"
+
+if [ "${repository_name}" == "gentoo" ]; then
+	ebegin "Fetching pre-generated metadata cache for ${repository_name}"
+	rsync -aq rsync://rsync.gentoo.org/gentoo-portage/metadata/md5-cache/ "${repository_path}"/metadata/md5-cache/
+	eend $?
+
+	ebegin "Updating metadata cache for ${repository_name}"
```
Comment (hasufell) [inline etc/portage/repo.postsync.d/sync_gentoo_cache:17]: we probably want thise OUTSIDE of the if-conditional, as in:

``` sh
if [ "${repository_name}" == "gentoo" ]; then
    ebegin "Fetching pre-generated metadata cache for ${repository_name}"
    rsync -aq rsync://rsync.gentoo.org/gentoo-portage/metadata/md5-cache/ "${repository_path}"/metadata/md5-cache/
    eend $?
fi

ebegin "Updating metadata cache for ${repository_name}"
egencache --jobs="${parallel_jobs}" --repo="${repository_name}" --update --update-use-local-desc
eend $?
```

File: etc/portage/repo.postsync.d/sync_gentoo_cache [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+source /etc/init.d/functions.sh
+source /etc/portage/util-functions.sh
+
+repository_name=${1}
+repository_path=${3}
+
+# Number of jobs for egencache, default is number or processors.
+parallel_jobs="$(nproc)"
+
+if [ "${repository_name}" == "gentoo" ]; then
```
Comment (hasufell) [inline etc/portage/repo.postsync.d/sync_gentoo_cache]: could you make this "real" bash syntax please? It's a bash script after all.

File: etc/portage/repo.postsync.d/sync_gentoo_cache (hunk 1 above)
Comment (redneb) [inline etc/portage/repo.postsync.d/sync_gentoo_cache:17]: Yes, that's what I did initially, but I think it does not make sense. Please read my justification in the third paragraph above. Of course, I can still move it outside if want.

**Comment by hasufell** [comment]
> why should it affect the behavior of the other repositories?

That's why it was initially an optional opt-in configuration. But @zx2c4 seemed to disagree. I guess I will make it optional again.

**Comment by redneb** [comment]
> That's why it was initially an optional opt-in configuration. But @zx2c4 seemed to disagree. I guess I will make it optional again

The way I see it, these configuration files are about enabling emerge sync use git. The problem is that the git repository we are syncing to does not have a pre-generated cache. So we should address that in order to have the same behavior as the traditional rsync based sync. But I feel that generating the cache for other repositories should be outside the scope of this project.

And again, if someone wants to to generated the cache for the other repositories, they may try to activate the already provided script:

```
chmod +x /etc/portage/repo.postsync.d/example
```

or:

```
cp -i /etc/portage/repo.postsync.d/example /etc/portage/repo.postsync.d/gen_cache
chmod +x /etc/portage/repo.postsync.d/gen_cache
```

**Comment by zx2c4** [comment]
> That's why it was initially an optional opt-in configuration. But @zx2c4 seemed to disagree. I guess I will make it optional again.

No. It either should be there, supporting all repos, and be enabled for all by default. Or it shouldn't be there at all. I'd prefer the latter in fact.

**Comment by hasufell** [comment]
https://bugs.gentoo.org/show_bug.cgi?id=558322 it is goint to be fixed in portage

Any particular reason anymore we want to move all the scripts to `repo.postsync.d` now?

**Comment by redneb** [comment]
There is still the issue of using the deprecated `PORTDIR` variable I mentioned in my first post: with `repo.postsync.d` we can get rid of it and also get rid of the `check-portdir.sh` auxiliary script.

**Comment by hasufell** [comment]
maybe makes sense

**Comment by hasufell** [comment]
we could add an example code (or even gist) in the README for people who want their overlays egencached as well

**Comment by redneb** [comment]
How about appending the following to the README. I can add it to the PR if you agree.

8<------------------------------------------------------------------------------

Note that it is generally a good idea to update the metadata-cache for all your other repositories as they typically don't come with a pregenerated cache. A good time to do that is when `emerge --sync` is ran. You can accomplish that by adding a hook to the `/etc/portage/postsync.d` directory, for example:

``` sh
#!/bin/bash

source /etc/init.d/functions.sh

# You may change this to only update cache of specific repositories.
repository_list=( $(portageq get_repos /) )

# Number of jobs for egencache, default is number or processors.
parallel_jobs="$(nproc)"

einfo "Updating metadata cache for repositories:"
for repo in "${repository_list[@]}"; do
    if [[ ${repo} != "gentoo" ]]; then
        ebegin "  ${repo}"
        egencache --jobs="${parallel_jobs}" --repo="${repo}" --update --update-use-local-desc
        eend $?
    fi
done
```

**Comment by hasufell** [comment]
I am currently setting up docker images for these configurations, so that I can automatically test it.

**Comment by hasufell** [comment]
> How about appending the following to the README

I think it is more consistent to do that in `repo.postsync.d` as well. Otherwise, it might cause weird behavior when you run `emaint sync <random-repo>` which will trigger egencache for all overlays then (I think).

**Comment by redneb** [comment]
> I think it is more consistent to do that in repo.postsync.d as well. Otherwise, it might cause weird behavior when you run emaint sync <random-repo> which will trigger egencache for all overlays then (I think).

I agree and I proposed a better solution above, I just thought you didn't like it. Anyways, here's that approach again with the accompanying text to be appended to the README:

8<------------------------------------------------------------------------------

Note that typically most repositories other than `gentoo` don't come with a pregenerated cache. It is a good idea to generate/update their cache every time they are synced. One way to do this, is by activating the `example` hook script that portage installs by default in `/etc/portage/postsync.d`:

```
cp -i /etc/portage/repo.postsync.d/example /etc/portage/repo.postsync.d/gen_cache
chmod +x /etc/portage/repo.postsync.d/gen_cache
```

**Comment by hasufell** [comment]
The new repository location has changed, see
https://gitweb.gentoo.org/data/gentoo-news.git/
and
https://github.com/hasufell/paludis-gentoo-git-config/commit/8e1d6c5ba50da89c6b4ad385708654d4a0558541

**Comment by hasufell** [comment]
here's my docker image btw https://github.com/hasufell/gentoo-docker-images/tree/portage/amd64-portage-git-tree

**Pull request merged**